core/src/main/java/org/infinispan/interceptors/impl/SimpleAsyncInvocationStage.java

```diff
@@ -69,12 +66,26 @@ public Object addCallback(InvocationContext ctx, VisitableCommand command,
          try {
             return function.apply(ctx, command, rv, throwable);
          } catch (Throwable t) {
-            return new SimpleAsyncInvocationStage(t);
+            return new ExceptionSyncInvocationStage(t);
          }
       }
       return new QueueAsyncInvocationStage(ctx, command, future, function);
    }
 
+   @Override
+   public Object thenReturn(InvocationContext ctx, VisitableCommand command, Object returnValue) {
+      if (future.isDone()) {
+         try {
+            future.getNow(null);
+            return returnValue;
+         } catch (Throwable t) {
```

Comment:
:+1:

Reply:
wait! why switch to `QueuedAsyncInvocationStage`?

Reply:
Because we expect further callbacks to be added, amortizing the queue in `QueuedAsyncInvocationStage`, instead of creating a `CompletableFuture` now and a `QueuedAsyncInvocationStage` later when we add another callback.

Reply:
makes sense :)